### ISSUE TITLE
key recovery

### ISSUE THREAD

**Issue by ooonea** [jedisct1/dnscrypt-server-docker#17]
Hello, after upgrading one of my servers, I was forced to reinstall the container and I rescheduled the keys with the method that waves in the guide (docker cp dnscrypt-server: / opt / dnscrypt-wrapper / etc / keys / tmp / keys - docker cp / tmp / keys dnscrypt-server: / opt / dnscrypt-wrapper / etc / keys). Copying is done correctly, but I can no longer use the server. In fact, dnscrypt-proxy tells me that the server has the certificate unusable (no longer recognizes it). What am I wrong?

**Comment by ooonea** [comment]
Solved by myself, I'm crazy because I'm not a docker, but I'm learning. I realized after hours that using the command 'docker cp / tmp / keys dnscrypt-server: / opt / dnscrypt-wrapper / etc / keys' the local 'keys' folder is copied to the 'keys' folder of dnscrypt-wrapper in the container . Then it then creates a 'keys' subfolder in the 'keys' folder and the correct files are not read when using the server startup. For the time being I have solved by issuing a copy command for each file of the local 'keys' folder to that of dnscrypt-wrapper. I've tried 'docker cp / tmp / keys / * dnscrypt-server: / opt / dnscrypt-wrapper / etc / keys', but it does not allow me to copy multiple files at once. I wonder: there is a mistake in your guide or I simply do not know how to use it. However, my servers are all healthy. Thank you for your patience and availability.

**Comment by jedisct1** [comment]
Good catch!

I think there is indeed an error in the guide. Maybe the existing directory should be removed first with `docker exec rm -rf /opt/dnscrypt-wrapper/etc/keys`.

**Comment by ooonea** [comment]
I think the guide should be corrected in another way. If you try to delete a folder when the container is off, do not delete it, because the folder is physically "off". If you try to delete it when the container is active, you will receive an error because the folder is in use (the keys are loaded). So I think the guide should be correct like this: 'docker cp / tmp / keys dnscrypt-server: / opt / dnscrypt-wrapper / etc /' (the 'keys' folder is overridden by the 'keys' folder in temp).
Otherwise, copy files one at a time: docker cp / tmp / keys / filename1 dnscrypt-server: / opt / dnscrypt-wrapper / etc / keys, docker cp / tmp / keys / filename2 ..., docker cp / tmp / keys / filename3 ... etc.
Also I tried to create that TLS termination with nginx to use servers with DNS over TLS, but it's really a mess, because I do not even know how to point to the container's localhost. Maybe in a million years I will succeed. I trust in your help;)

**Comment by mibere** [comment]
Regarding updating a existing/running server: the guide https://github.com/jedisct1/dnscrypt-proxy/wiki/Manually-updating-your-DNSCrypt-server already contains

`docker exec dnscrypt-server-new rm -rf /opt/dnscrypt-wrapper/etc/keys/short-term`

I followed this guide for my [dnscrypt.me](http://dnscrypt.me/) several times, without any problems so far.

**Comment by ooonea** [comment]
mibere, The problem is not that, I deleted the short-term folder. The problem is that the saved keys were copied to a subfolder of the 'keys' folder in '/opt /dnscrypt-wrapper/etc/keys'. So the operation is wrong upstream. On https://github.com/jedisct1/dnscrypt-proxy/wiki/Manually-updating-your-DNSCrypt-server.
In fact it is: 'docker cp ~ / keys /. dnscrypt-new-server: / opt / dnscrypt-wrapper / etc / keys'
I really think that '.' in .. 'cp ~ / keys /.' make the difference.
He says to copy the contents of the 'keys' folder and not the folder itself. This is omitted in the guide I followed instead: https://github.com/jedisct1/dnscrypt-proxy/wiki/How-to-setup-your-own-DNSCrypt-server-in-less-than-10- minutes
Now I edited the wiki with the command that works.
